lib/chef/cookbook_manifest.rb

```diff
@@ -0,0 +1,275 @@
+
+    def_delegator :@cookbook_version, :root_paths
+    def_delegator :@cookbook_version, :segment_filenames
+    def_delegator :@cookbook_version, :name
+    def_delegator :@cookbook_version, :metadata
+    def_delegator :@cookbook_version, :full_name
+    def_delegator :@cookbook_version, :version
+    def_delegator :@cookbook_version, :frozen_version?
+
+    # Create a new CookbookManifest object for the given `cookbook_version`.
+    # You can subsequently call #to_hash to get a Hash representation of the
+    # cookbook_version in the "manifest" format, or #to_json to get a JSON
+    # representation of the cookbook_version.
+    #
+    # The inferface for this behavior is expected to change as we implement new
+    # manifest formats. The entire class should be considered a private API for
+    # now.
+    #
+    # @api private
+    # @param policy_mode [Boolean] whether to convert cookbooks to Hash/JSON in
+    #   the format used by the `cookbook_artifacts` endpoint (for policyfiles).
+    #   Setting this option also changes the behavior of #save_url and
+    #   #force_save_url such that CookbookVersions will be uploaded to the new
+    #   `cookbook_artifacts` API. This endpoint is currently under active
+    #   development and the format is expected to change frequently, therefore
+    #   the result of #manifest, #to_hash, and #to_json will not be stable when
+    #   `policy_mode` is enabled.
+    def initialize(cookbook_version, policy_mode: false)
+      @cookbook_version = cookbook_version
+      @policy_mode = !!policy_mode
+
+      reset!
+    end
+
+    # Resets all lazily computed values.
+    def reset!
+      @manifest = nil
+      @checksums = nil
+      @manifest_records_by_path = nil
+      true
+    end
+
+    # Returns a 'manifest' data structure that can be uploaded to a Chef
+    # Server.
+    #
+    # The format is as follows:
+    #
+    #     {
+    #       :cookbook_name  => name,            # String
+    #       :metadata       => metadata,        # Chef::Cookbook::Metadata
+    #       :version        => version,         # Chef::Version
+    #       :name           => full_name,       # String of "#{name}-#{version}"
+    #
+    #       :recipes        => Array<FileSpec>,
+    #       :definitions    => Array<FileSpec>,
+    #       :libraries      => Array<FileSpec>,
+    #       :attributes     => Array<FileSpec>,
+    #       :files          => Array<FileSpec>,
+    #       :templates      => Array<FileSpec>,
+    #       :resources      => Array<FileSpec>,
+    #       :providers      => Array<FileSpec>,
+    #       :root_files     => Array<FileSpec>
+    #     }
+    #
+    # Where a `FileSpec` is a Hash of the form:
+    #
+    #     {
+    #       :name         => file_name,
+    #       :path         => path,
+    #       :checksum     => csum,
+    #       :specificity  => specificity
+    #     }
+    #
+    def manifest
+      @manifest || generate_manifest
+      @manifest
```

Comment:
Technically, yeah, since the manifest is the return value of `generate_manifest`. On the other hand this is consistent with the other two methods that rely on the side effects of `generate_manifest`. I could go either way, but I think this way slightly emphasizes that `generate_manifest` is expected to mutate the object and set the instance variable.